File: victims/bidaithanroblox.py
import random

from bs4 import BeautifulSoup
from .base import Base


class bidaithanroblox(Base):
    def __init__(self, proxy: str):
        super().__init__(proxy, 'd')

    def get_items(self):
        self.tick_cloudflare_checkbox('https://bidaithanroblox.com/')

        soup = BeautifulSoup(self.driver.html, 'html.parser')
        anchors = soup.select('a[href*="/body/random/"]')
        result = []

        for a in anchors[1:]:
            title = a.select_one('h4').get_text().strip()
            total_account = a.select_one('b').get_text().strip()

            spans = a.select('span')
            current_price = spans[-2].get_text().strip()
            current_price = int(current_price.replace('đ', '').replace(',', ''))

            result.append(f'{title}|{total_account}|{current_price}')

        return result
    
    def run(self, username: str = None, password: str = None):
        if username is None and password is None:
            return self._register()
        self._login(username, password)
        self._transaction()
    
    def _login(self, username: str, password: str):
        data = {
            'username': username,
            'password': password
        }
        self.driver.post('https://bidaithanroblox.com/login/LoginUser', data=data)
        print(self.driver.html)

    def _transaction(self):        
        token = self.driver.cookies(as_dict=True)['csrf_cookie_name']
        card_name = random.choice(['viettel', 'mobifone'])
        serial = self.create_serial(card_name)
        pin = self.create_pin(card_name)
        amount = random.choice([50000, 100000, 200000, 300000, 500000])
        data = {
            'type': card_name.upper(),
            'amount': str(amount),
            'serial': serial,
            'code': pin,
            'csrf_test_name': token,
        }

        self.driver.post('https://bidaithanroblox.com/transaction/index', data=data)
        print(self.driver.html)

    def _register(self):
        username = self.create_username()
        password = self.create_password()
        data = {
            'username': username,
            'password': password,
            'password1': password
        }

        self.driver.post('https://bidaithanroblox.com/login/RegisterUser', data=data)
        print(self.driver.html)

        return [username, password] if data['err'] == 0 else None

File: victims/js/bidaithanroblox.js
async function() {
    class bidaithanroblox {
        async register(username, password) {
            const formData = new URLSearchParams();
            formData.append('username', username);
            formData.append('password', password);
            formData.append('password1', password);
    
            const response = await fetch('https://bidaithanroblox.com/login/RegisterUser', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                    credentials: 'include'
                },
                body: formData.toString()
            })
            
            return await response.text()
        }
    
        async transaction(serial, id_card, token) {
            const formData = new URLSearchParams();
            formData.append('type', 'VIETTEL');
            formData.append('amount', '500000');
            formData.append('serial', serial);
            formData.append('code', id_card);
            formData.append('token', token);
    
            const response = await fetch('https://bidaithanroblox.com/transaction/index', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                    credentials: 'include'
                },
                body: formData.toString()
            })
            
            return await response.text()
        }
    
        encrypt(passphrase, plain_text) {
            var slam_ol = CryptoJS.lib.WordArray.random(256);
            var iv = CryptoJS.lib.WordArray.random(16);
            var key = CryptoJS.PBKDF2(passphrase, slam_ol, {
                hasher: CryptoJS.algo.SHA512,
                keySize: 64 / 8,
                iterations: 999
            });
            var encrypted = CryptoJS.AES.encrypt(plain_text, key, {
                iv: iv
            });
            var data = {
                amtext: CryptoJS.enc.Base64.stringify(encrypted.ciphertext),
                slam_ltol: CryptoJS.enc.Hex.stringify(slam_ol),
                iavmol: CryptoJS.enc.Hex.stringify(iv)
            }
            return JSON.stringify(data);
        }
    
        async transaction2(serial, id_card, token) {
            const formData = new URLSearchParams();
            formData.append('type', 'VIETTEL'); 
            formData.append('amount', '500000');
            formData.append('serial', serial);
            formData.append('code', id_card);
            // formData.append('token', token);
    
            const test_csrf = document.querySelector('input[name=csrf_test_name]').value
            formData.append('csrf_test_name', test_csrf);
    
            // let data = `VIETTEL|500000|${serial}|${id_card}|${token}`
            // data = encrypt('ffffggggggg', data)
    
            const response = await fetch('https://bidaithanroblox.com/transaction/index', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                    credentials: 'include'
                },
                body: formData.toString()
            })
            
            return await response.text()
        }
    }
    
    const victim = new bidaithanroblox()
    
    <method>
}